File: boj_codes/Python3/1629.py
# https://www.acmicpc.net/problem/1629
# 20201030

import sys


def exponent_remainder(base, exponent, div, original_base):
    if exponent == 0:
        return 1
    elif exponent == 1:
        base %= div
        return base
    else:
        base = exponent_remainder(base, exponent // 2, div, original_base)
        if exponent % 2 == 0:
            base *= base
        elif exponent % 2 == 1:
            base *= original_base * base

        base %= div
        return base


base, exponent, div = map(int, sys.stdin.readline().split())

print(exponent_remainder(base, exponent, div, base))

File: old_contents/boj_codes/C++17/1629.cpp
// https://www.acmicpc.net/problem/1629
// First Written : 20210101
// Last Modified : 20210101

#include <cstdio>

long long exponent(long long base, long long exp, long long div){
    if(exp == 0)
        return 1;
    else if(exp == 1){
        base %= div;
        return base;
    }
    else{
        long long old_base = base;
        base = exponent(base, exp / 2, div);

        if(exp % 2 == 0){
            base *= base;
        }
        else{
            base = (base * base) % div;
            base *= old_base;
        }

        base %= div;
        return base;
    }
        
}

int main(){
    long long base, exp, div;
    long long result = 1;
    scanf("%lld %lld %lld", &base, &exp, &div);

    printf("%lld", exponent(base, exp, div));

    return 0;
}